node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/SpareCapacityMaintainer.java

```diff
@@ -0,0 +1,346 @@
+
+        CapacityChecker capacityChecker = new CapacityChecker(nodeRepository());
+
+        List<Node> overcommittedHosts = capacityChecker.findOvercommittedHosts();
+        if (overcommittedHosts.size() != 0) {
+            log.log(Level.WARNING, String.format("%d nodes are overcommitted! [ %s ]",
+                                                 overcommittedHosts.size(),
+                                                 overcommittedHosts.stream().map(Node::hostname).collect(Collectors.joining(", "))));
+        }
+        metric.set("overcommittedHosts", overcommittedHosts.size(), null);
+
+        Optional<CapacityChecker.HostFailurePath> failurePath = capacityChecker.worstCaseHostLossLeadingToFailure();
+        if (failurePath.isPresent()) {
+            int spareHostCapacity = failurePath.get().hostsCausingFailure.size() - 1;
+            if (spareHostCapacity == 0) {
+                Move move = findMitigatingMove(failurePath.get());
+                if (moving(move)) {
+                    // We succeeded or are in the process of taking a step to mitigate.
+                    // Report with the assumption this will eventually succeed to avoid alerting before we're stuck
+                    spareHostCapacity++;
+                }
+            }
+            metric.set("spareHostCapacity", spareHostCapacity, null);
+        }
+    }
+
+    private boolean moving(Move move) {
+        if (move.isEmpty()) return false;
+        if (move.node().allocation().get().membership().retired()) return true; // Move already in progress
+        return move.execute(false, Agent.SpareCapacityMaintainer, deployer, metric, nodeRepository());
+    }
+
+    private Move findMitigatingMove(CapacityChecker.HostFailurePath failurePath) {
+        Optional<Node> nodeWhichCantMove = failurePath.failureReason.tenant;
+        if (nodeWhichCantMove.isEmpty()) return Move.empty();
+
+        Node node = nodeWhichCantMove.get();
+        NodeList allNodes = nodeRepository().list();
+        // Allocation will assign the two most empty nodes as "spares", which will not be allocated on
+        // unless needed for node failing. Our goal here is to make room on these spares for the given node
+        HostCapacity hostCapacity = new HostCapacity(allNodes, nodeRepository().resourcesCalculator());
+        Set<Node> spareHosts = hostCapacity.findSpareHosts(allNodes.hosts().satisfies(node.resources()).asList(), 2);
+        List<Node> hosts = allNodes.hosts().except(spareHosts).asList();
+
+        CapacitySolver capacitySolver = new CapacitySolver(hostCapacity, maxIterations);
+        List<Move> shortestMitigation = null;
+        for (Node spareHost : spareHosts) {
+            List<Move> mitigation = capacitySolver.makeRoomFor(node, spareHost, hosts, List.of(), List.of());
+            if (mitigation == null) continue;
+            if (shortestMitigation == null || shortestMitigation.size() > mitigation.size())
```

Comment:
It's not obvious to me that this return `null`/empty list always ends being correct, why not always return non-null?

Also, add a check here to ignore empty mitigation if we already have a `shortestMitigation` candidate, after all doing something that takes longer should be better than doing nothing?

Reply:
Empty list means the solution needs no moves.
null means there are no solutions.

I don't understand the second paragraph - if it's null we continue because we found no solution. Otherwise we compare it to the current best, if any.

Reply:
If `makeRoomFor()` returns empty list, it will be set as the `shortestMitigation` since `shortestMitigation.size() > mitigation.size()` == `shortestMitigation.size() > 0`, but then 2 lines below, if `shortestMitigation.isEmpty()` we return `Move.empty()`. So wouldn't it be better to change this if to
```
if (shortestMitigation == null || (shortestMitigation.size() > mitigation.size() && !mitigation.isEmpty()))
```
?

Reply:
If the shortest mitigation is empty it means that the situation can be mitigated in 0 moves, i.e it is already fine.
This is an inconsistency that will only arise if some change happened to mitigate it between when the nodes were read by CapacityChecker and by this, or if either the CapacityChecker or this has a bug. I don't think we should prefer an unnecessary non-empty mitigation in this case, but arguably we could to signal it upwards such that we don't emit the alert in this case (and perhaps also log it). I'm uncertain on whether it's worth the extra complication.